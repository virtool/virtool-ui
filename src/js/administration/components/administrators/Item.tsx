import React from "react";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../../app/theme";
import { BoxSpaced, Icon, InitialIcon } from "../../../base";
import { User } from "../../../users/types";
import { useSetAdministratorRole } from "../../queries";
import { AdministratorRoles } from "../../types";
import { RoleSelect } from "./RoleSelect";

const StyledAdministrator = styled(BoxSpaced)`
    display: flex;
    align-items: center;

    svg {
        margin-right: 10px;
    }

    button {
        margin-left: auto;
    }

    button:last-child {
        margin-left: 10px;
    }
`;

const UserHandle = styled.span`
    font-weight: ${getFontWeight("thick")};
    font-size: ${getFontSize("lg")};
`;

const InlineRoleSelect = styled(RoleSelect)`
    width: 200px;
`;

type AdministratorItemProps = {
    user: User;
    roles: Array<AdministratorRoles>;
};

export const AdministratorItem = ({ user, roles }: AdministratorItemProps) => {
    const editMutation = useSetAdministratorRole();
    const onChange = (value: AdministratorRoles) => {
        editMutation.mutate({ user_id: user.id, role: value });
    };

    return (
        <StyledAdministrator key={user.id}>
            <InitialIcon handle={user.handle} size="lg" />
            <UserHandle>{user.handle}</UserHandle>
            <InlineRoleSelect value={user.administrator_role} roles={roles} onChange={onChange} />
            <Icon name="trash" color="red" onClick={() => onChange(null)} aria-label="Remove administrator role" />
        </StyledAdministrator>
    );
};
