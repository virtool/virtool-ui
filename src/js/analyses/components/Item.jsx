import React from "react";
import { connect } from "react-redux";
import { Link, useRouteMatch } from "react-router-dom";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../app/theme";
import { Attribution, BoxSpaced, Icon, SlashList } from "../../base";
import { getCanModify } from "../../samples/selectors";
import { getWorkflowDisplayName } from "../../utils/utils";
import { removeAnalysis } from "../actions";
import { AnalysisItemRightIcon } from "./RightIcon";

const StyledAnalysisItem = styled(BoxSpaced)`
    color: ${props => props.theme.color.greyDarkest};

    &:hover {
        ${props => (props.ready ? "background-color: lightgrey;" : "")};
    }
`;

const AnalysisItemTag = styled.span`
    align-items: center;
    display: inline-flex;
    margin-right: 15px;

    ${SlashList} {
        margin: 0;
    }

    i {
        margin-right: 5px;
    }
`;

const AnalysisItemTags = styled.div`
    align-items: center;
    display: flex;
    margin-top: 10px;
`;

const AnalysisItemTop = styled.div`
    align-items: center;
    display: flex;
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};
    justify-content: space-between;

    a {
        font-weight: ${getFontWeight("thick")};
    }
`;

function AnalysisItem({ canModify, created_at, id, index, ready, reference, subtractions, user, workflow, onRemove }) {
    const sampleId = useRouteMatch().params.sampleId;

    return (
        <StyledAnalysisItem>
            <AnalysisItemTop>
                <Link to={`/samples/${sampleId}/analyses/${id}`}>{getWorkflowDisplayName(workflow)}</Link>
                <AnalysisItemRightIcon canModify={canModify} onRemove={onRemove} ready={ready} />
            </AnalysisItemTop>
            <Attribution user={user.handle} time={created_at} />
            <AnalysisItemTags>
                <AnalysisItemTag key="reference">
                    <Icon name="equals" />
                    <SlashList>
                        <li>
                            <Link to={`/refs/${reference.id}`}>{reference.name}</Link>
                        </li>
                        <li>
                            <Link to={`/refs/${reference.id}/indexes/${index.id}`}>Index {index.version}</Link>
                        </li>
                    </SlashList>
                </AnalysisItemTag>
                {subtractions.map(subtraction => (
                    <AnalysisItemTag key={subtraction.id}>
                        <Icon name="not-equal" />
                        <Link to={`/subtractions/${subtraction.id}`}>{subtraction.name}</Link>
                    </AnalysisItemTag>
                ))}
            </AnalysisItemTags>
        </StyledAnalysisItem>
    );
}

export function mapStateToProps(state) {
    return {
        canModify: getCanModify(state)
    };
}

function mapDispatchToProps(dispatch, ownProps) {
    return {
        onRemove: () => {
            dispatch(removeAnalysis(ownProps.id));
        }
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(AnalysisItem);
