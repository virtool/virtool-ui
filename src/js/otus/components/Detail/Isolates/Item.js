import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { getActiveShadow } from "../../../../app/theme";
import { BoxGroupSection, Icon } from "../../../../base";
import { formatIsolateName } from "../../../../utils/utils";

const StyledIsolateItem = styled(BoxGroupSection)`
    align-items: center;
    border: none;
    box-shadow: ${getActiveShadow};
    display: flex;

    & > span {
        overflow: hidden;
        text-overflow: ellipsis;
        white-space: nowrap;
    }

    i.fas {
        margin-left: auto;
    }
`;

export const IsolateItem = props => {
    const handleSelectIsolate = useCallback(() => {
        props.onClick(props.id);
    }, [props.id]);

    return (
        <StyledIsolateItem active={props.active} onClick={handleSelectIsolate}>
            <span>{formatIsolateName(props)}</span>
            {props.default && props.dataType !== "barcode" && <Icon name="star" />}
        </StyledIsolateItem>
    );
};

export const mapStateToProps = state => ({
    dataType: state.references.detail.data_type
});

export default connect(mapStateToProps)(IsolateItem);
