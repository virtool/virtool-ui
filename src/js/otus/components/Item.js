import React from "react";
import { get } from "lodash-es";
import styled from "styled-components";
import { connect } from "react-redux";
import { getFontSize, getFontWeight } from "../../app/theme";
import { LinkBox, Icon } from "../../base";

const OTUItemAbbreviation = styled.span`
    margin-left: auto;
`;

const OTUItemName = styled.strong`
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};
`;

const StyledOTUItem = styled(LinkBox)`
    align-items: center;
    display: flex;
`;

export const OTUItem = ({ abbreviation, id, name, refId, verified }) => (
    <StyledOTUItem key={id} to={`/refs/${refId}/otus/${id}`}>
        <OTUItemName>{name}</OTUItemName>
        <OTUItemAbbreviation>{abbreviation}</OTUItemAbbreviation>
        {verified ? null : <Icon name="tag" tip="This OTU is unverified" />}
    </StyledOTUItem>
);

export const mapStateToProps = (state, props) => {
    const { abbreviation, id, name, verified } = get(state, ["otus", "documents", props.index]);
    return {
        abbreviation,
        id,
        name,
        verified,
        refId: state.references.detail.id
    };
};

export default connect(mapStateToProps)(OTUItem);
