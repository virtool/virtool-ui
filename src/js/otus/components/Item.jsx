import { get } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import styled from "styled-components";
import { getFontSize, getFontWeight } from "../../app/theme";
import { Box, Icon } from "../../base";

const OTUItemName = styled(Link)`
    font-size: ${getFontSize("lg")};
    font-weight: ${getFontWeight("thick")};
`;

const OtuItemUnverified = styled.span`
    align-items: center;
    display: flex;
    gap: 5px;
    justify-content: flex-end;
`;

const StyledOTUItem = styled(Box)`
    align-items: center;
    display: grid;
    grid-template-columns: 5fr 2fr 1fr;
`;

export function OTUItem({ abbreviation, id, name, refId, verified }) {
    return (
        <StyledOTUItem key={id}>
            <OTUItemName to={`/refs/${refId}/otus/${id}`}>{name}</OTUItemName>
            <span>{abbreviation}</span>
            {verified || (
                <OtuItemUnverified>
                    <Icon name="tag" />
                    Unverified
                </OtuItemUnverified>
            )}
        </StyledOTUItem>
    );
}

export function mapStateToProps(state, props) {
    const { abbreviation, id, name, verified } = get(state, ["otus", "documents", props.index]);

    return {
        abbreviation,
        id,
        name,
        verified,
        refId: state.references.detail.id,
    };
}

export default connect(mapStateToProps)(OTUItem);
