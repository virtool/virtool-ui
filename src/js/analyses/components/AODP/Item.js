import React, { useCallback } from "react";
import styled from "styled-components";
import { connect } from "react-redux";
import { Badge } from "../../../base";

import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";

const StyledAODPItem = styled(AnalysisViewerItem)`
    display: flex;
    justify-content: space-between;
`;

export const AODPItem = ({ active, hits, id, name, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(id), [id]);

    return (
        <StyledAODPItem active={active} style={style} onClick={handleClick}>
            <strong>{name}</strong>
            <Badge>{hits}</Badge>
        </StyledAODPItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).id;
    const matches = getMatches(state);
    const match = matches[ownProps.index];

    const { id, identities, name } = match;
    return { id, name, hits: identities.length, active: activeId === id };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AODPItem);
