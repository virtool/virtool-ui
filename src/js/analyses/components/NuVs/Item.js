import { filter } from "lodash-es";
import numbro from "numbro";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { Badge } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";
import { NuVsValues } from "./Values";

const calculateAnnotatedOrfCount = orfs => filter(orfs, orf => orf.hits.length).length;

const NuVsItemHeader = styled.div`
    align-items: center;
    display: flex;
    justify-content: space-between;
`;

const StyledNuVsItem = styled(AnalysisViewerItem)`
    border-bottom: none;
    border-left: none;
    border-radius: 0;
    margin: 0;
`;

export const NuVsItem = ({ active, e, orfs, sequence, sequenceIndex, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(sequenceIndex), [sequenceIndex]);

    return (
        <StyledNuVsItem active={active} onClick={handleClick} style={style}>
            <NuVsItemHeader>
                <strong>Sequence {sequenceIndex}</strong>
                <Badge>{sequence.length}</Badge>
            </NuVsItemHeader>
            <NuVsValues e={numbro(e).format()} orfCount={calculateAnnotatedOrfCount(orfs)} />
        </StyledNuVsItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).index;
    const { e, index, orfs, sequence } = getMatches(state)[ownProps.index];
    return { e, orfs, sequence, sequenceIndex: index, active: activeId === index };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NuVsItem);
