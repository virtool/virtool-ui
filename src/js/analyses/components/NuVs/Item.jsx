import numbro from "numbro";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { Badge } from "../../../base";
import { setActiveHitId } from "../../actions";
import { getActiveHit, getMatches } from "../../selectors";
import { AnalysisViewerItem } from "../Viewer/Item";
import { NuVsValues } from "./Values";

const NuVsItemHeader = styled.div`
    align-items: center;
    display: flex;
    justify-content: space-between;
`;

const StyledNuVsItem = styled(AnalysisViewerItem)`
    border-bottom: none;
    border-left: none;
    border-radius: 0;
    margin: 0;
`;

export const NuVsItem = ({ active, e, orfCount, sequence, sequenceIndex, style, onSetActiveId }) => {
    const handleClick = useCallback(() => onSetActiveId(sequenceIndex), [sequenceIndex]);
    return (
        <StyledNuVsItem active={active} onClick={handleClick} style={style}>
            <NuVsItemHeader>
                <strong>Sequence {sequenceIndex}</strong>
                <Badge>{sequence.length}</Badge>
            </NuVsItemHeader>
            <NuVsValues e={numbro(e).format()} orfCount={orfCount} />
        </StyledNuVsItem>
    );
};

const mapStateToProps = (state, ownProps) => {
    const activeId = getActiveHit(state).index;
    const { e, index, annotatedOrfCount, sequence } = getMatches(state)[ownProps.index];
    return { e, orfCount: annotatedOrfCount, sequence, sequenceIndex: index, active: activeId === index };
};

const mapDispatchToProps = dispatch => ({
    onSetActiveId: id => {
        dispatch(setActiveHitId(id));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NuVsItem);
