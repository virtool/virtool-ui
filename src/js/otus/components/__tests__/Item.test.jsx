import { createBrowserHistory } from "history";
import React from "react";
import { beforeEach, describe, expect, it } from "vitest";
import { renderWithRouter } from "../../../../tests/setupTests";
import { mapStateToProps, OTUItem } from "../Item";

describe("<OTUItem />", () => {
    let props;

    beforeEach(() => {
        props = {
            abbreviation: "FB",
            id: "foo",
            name: "Foo",
            verified: true,
            refId: "baz",
        };
    });

    it("should render when [verified=true]", () => {
        const history = createBrowserHistory();

        renderWithRouter(<OTUItem {...props} />, {}, history);

        const link = screen.getByRole("link");

        expect(link).toHaveTextContent("Foo");
        expect(link).toHaveAttribute("href", "/refs/baz/otus/foo");
        expect(screen.queryByText("Unverified")).not.toBeInTheDocument();
    });

    it("should render when [verified=false]", () => {
        const history = createBrowserHistory();

        renderWithRouter(<OTUItem {...props} verified={false} />, {}, history);

        const link = screen.getByRole("link");

        expect(link).toHaveTextContent("Foo");
        expect(link).toHaveAttribute("href", "/refs/baz/otus/foo");

        expect(screen.getByText("Unverified")).toBeInTheDocument();
    });
});

describe("mapStateToProps()", () => {
    it("should return props given state", () => {
        const state = {
            otus: {
                documents: [
                    {
                        id: "foo",
                        name: "Foo",
                        abbreviation: "FO",
                        verified: true,
                    },
                    {
                        id: "bar",
                        name: "Bar",
                        abbreviation: "BR",
                        verified: true,
                    },
                    {
                        id: "baz",
                        name: "Baz",
                        abbreviation: "BZ",
                        verified: true,
                    },
                ],
            },
            references: {
                detail: {
                    id: "ref",
                },
            },
        };
        const result = mapStateToProps(state, { index: 1 });
        expect(result).toEqual({
            id: "bar",
            name: "Bar",
            abbreviation: "BR",
            verified: true,
            refId: "ref",
        });
    });
});
