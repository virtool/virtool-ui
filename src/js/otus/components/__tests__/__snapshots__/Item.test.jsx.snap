// Vitest Snapshot v1

exports[`<OTUItem /> > should render when [verified=false] 1`] = `
<Styled(BoxLink)
  key="foo"
  to="/refs/baz/otus/foo"
>
  <styled.strong>
    Foo
  </styled.strong>
  <styled.span>
    FB
  </styled.span>
  <Icon
    name="tag"
    tip="This OTU is unverified"
  />
</Styled(BoxLink)>
`;

exports[`<OTUItem /> > should render when [verified=true] 1`] = `
<Styled(BoxLink)
  key="foo"
  to="/refs/baz/otus/foo"
>
  <styled.strong>
    Foo
  </styled.strong>
  <styled.span>
    FB
  </styled.span>
</Styled(BoxLink)>
`;
